src/uscheduleconf.1.in
.TH uscheduleconf 1 \*(VE \*(PA      
.hy 0
.URL "Uwe Ohse" ../uwe.html
.br
.URL "The uschedule package" ../uschedule.html
.paragraph
.SH NAME
uscheduleconf - configure a scheduling service
.LINE
.section SYNOPSIS
\fBuscheduleconf\fR DIR ACCT LOGACCT [JOBDIR [LOGDIR]]
.section DESCRIPTION
\fBuscheduleconf\fR creates a svscan service directory \fIDIR\fR, which
starts a service to uschedule for the user \fIACCT\fR. The jobs will be read
from \fIJOBDIR\fR (or ~\fIACCT\fR/.uschedule, if \fIJOBDIR\fR is not given
or only contains a minus character).
.paragraph
Logging information will be written to \fILOGDIR\fR 
(or ~\fIACCT\fR/.uschedule/log, if \fILOGDIR\fR is not given), using the
the account \fILOGACCT\fR.
.paragraph
Version before 0.7.0 allowed to give "-" for ACCT to use the current user
and "-" for LOGACCT to use ACCT for LOGACCT, but the code turned out to
be buggy. See the NEWS file of the uschedule package for more information.
.section "OPTIONS"
.options
.option "-n, --no-user-change"
Do not change file ownership and do not switch user ids.
.optionparagraph
The default is to make ACCT and LOGACCT the owners of JOBDIR, LOGDIR and
the file therein, and to switch to ACCT and LOGACCT before running 
uscheduled and the logging program.
.optionparagraph
With this option the caller will be the owner of all files created and 
the programs will run with all the rights they inherit.
.optionparagraph
This option was added in version 0.7.0.
.endoptions
.section "ENVIRONMENT"
In \fIJOBDIR\fR/env a number of environment variables are set (one file
per variable). HOME, SHELL, USER and LOGNAME are set to values taken 
from the system password database.
.br
The PATH variable is set to /command:/usr/local/bin:/usr/bin/:/bin 
for any user not having an uid of 0. For users with uid 0 
/usr/local/sbin:/usr/sbin:/sbin will be appended. 
.paragraph
All these variables may be changed.
.paragraph
.section "SECURITY"
Keep the following rules in mind:
.list
.entry
Do not ever create a service directory \fIDIR\fR in an insecure place.
Nobody but you - the system administrator - must be able to write to it.
Especially do not ever create this directory in the users $HOME. You
would be giving away instant root access!
.entry
Do not ever create \fIJOBDIR\fR and \fILOGDIR\fR in an insecure place.
These directories and the content must only be writable by the user
for which the service is run.
.entry
Be careful about \fILOGACCT\fR. If you set up a scheduling service for
a user then \fILOGACCT\fR should be this user. If you are setting up
a scheduling service for a system account then you may want to use
a different user for logging purposes (while this is local policy i can't
think of any reason why the logging should run as root).
.entry
Resource limits may be configured in \fIDIR/run\fR and \fIDIR/log/run\fR.
The defaults are most possibly not right for every system. Keep in mind
that perl eats lots of resources and that a multilog processor may use
perl ...
.endlist
.section "EXAMPLES"
.subsection "Setup a scheduler for system services"
Create a uschedule service for root, with the service
directory /etc/root-schedule and the logging done to 
/var/log/root-schedule under the "misclog" account.
.paragraph
.nf
.HTML <pre>
    uscheduleconf /etc/root-schedule root misclog \\
             /etc/root-schedule/jobs /var/log/root-schedule
.HTML </pre>
.fi
.subsection "Setup a scheduler for a user"
Create a schedule service for uwe, having the trusted 
commands in /etc/uwe-schedule and any parts uwe can
change in ~uwe/.uschedule:
.paragraph
.nf
.HTML <pre>
    uscheduleconf /etc/uwe-schedule uwe uwe \\
             ~uwe/.uschedule ~uwe/.uschedule/logs
.HTML </pre>
.fi
.section AUTHOR
Uwe Ohse, uwe@ohse.de
.section "SEE ALSO"
.manlink uschedule.html uschedule(1) ,
.manlink uschedulecmd.html uschedulecmd(1) ,
.manlink intro.html uschedule_intro(7) .
.HTML <!--
.paragraph
The homepage may be more up-to-date, see
.br
http://www.ohse.de/uwe/uschedule.html.
.HTML -->


src/NEWS
Version 0.7.1, August 2004:

* fixed uschedulelist --short (-s), which used to display the "next" 
  instead of the "last" time.
* "uschedule --every X" (-e) repeats a job with a given interval X.
* The new uscheduleruntimelimit tool allows to limit the run time of a job.
* uscheduleconf makes base directory sticky for daemontools <0.75.
* uschedulelist --show-command (-S) shows the command in addition to the 
  usual list output.
* uschedule accepts + as synonym for / in TIMESPEC.

Version 0.7.0, March 2003:

* implemented -. option ("use . instead of $HOME").
* bugfix, "-" as username or logusername was mistreated.

  *** Impact: the user may have got root rights. ***

  Note: older versions may have left a multilog running as root instead
  of the user. Impact: the user may have got root rights.
  Fix: Go though /service/*uschedule*/log/run and add 
    setuidgid "username" \
  after the line containing the softlimit call and before the line 
  containing run. Afterwards the file should look like this:
    #! /bin/sh
    exec 2>&1
    cd '...SOMEWHERE...' || exit 1
    exec \
    softlimit -m 8000000 -o 400 -p 40 \
    setuidgid USERNAME \
    ./run

  The code containing the bug has been rewritten and simplified.

* uscheduleconf now has a --no-user-change option.

Version 0.6.1, October 2002:

* installed manual pages again.

Version 0.6.0, October 2002:

* uschedulecmd used to steal and misinterpret options given to the child 
  command. Now all programs stop parsing looking for options after the
  first non-option argument.
* implemented start/end times for jobs. See the uschedule manual page
  for more information.
* uschedulelist now prints all month in the 1..12 range, not 0..11 anymore.

Version 0.5.8, November 2001:

* internal changes to get rid of some warnings.
* the same format is used whereever a time is printed.

Version 0.5.7, November 2001:

* uscheduleconf created completely broken configuration if a log account
  different from the user account was given.

Version 0.5.6, October 2001:

* uscheduleconf created buggy call to multilog.

Version 0.5.5, October 2001:

* fixed another self check problem.

Version 0.5.4, October 2001:

* fixed a number of documentation bugs.
* uscheduleconf now starts uscheduled instead of the nonexistant scheduled.
* self check now works if the machine has a time zone different from UTC
  or the local file system has a block size different from 8k.

Version 0.5.3, September 2001:

* first public release.

Version 0.5.1 to 0.5.2, September 2001:

* internal release.

Version 0.5.0, August 2001:

* first internal release.


src/uscheduleconf.c
#include <sys/types.h>
#include <pwd.h>
#include <unistd.h>
#include "svscan_conf.h"
#include "bailout.h"
#include "str.h"
#include "env.h"
#include "scheduled.h"
#include "uogetopt.h"

static const char *dir;
static int rootuid=0;
static int rootgid=0;
static int useruid;
static int usergid;
static int loguid;
static int loggid;

static int o_user_change=1;
static void xowner(uid_t u, gid_t g) { if (o_user_change) owner(u,g); }
static void xsetuidgid(const char *s)
{ if (!o_user_change) return;
  outs("setuidgid "); outs(s); outs(" \\\n");
}

static void mp(const char *s, int mode, uid_t u, gid_t g)
{
	static stralloc d=STRALLOC_INIT;
	if (!stralloc_copys(&d,dir)) oom();
	if (!stralloc_cats(&d,"/")) oom();
	if (!stralloc_cats(&d,s)) oom();
	if (!stralloc_0(&d)) oom();
	make_dir(d.s);
	xowner(u,g);
	perm(mode);
}

static uogetopt2 myopts[]={
{'n',"no-user-change",uogo_flag,UOGO_NOARG, &o_user_change, 0 , 
"Don't change file ownership, don't switch user ids.",
"The default is to make ACCT and LOGACCT the owners "
"of many files, and to switch to ACCT and LOGACCT "
"before running uscheduled and the logging program. "
"With this option the caller will be the owner of "
"all files created and the programs will run with "
"all the rights they inherit.",0},
{0,0,0,0,0,0,0,0,0}
};

static void simple_env(const char *f, const char *val)
{
	start_file(f);
	outs(val); outs("\n");
	finish_file();
	xowner(useruid,usergid);
	perm(0700);
}
static uogetopt_env myoptenv={
"uscheduleconf",PACKAGE,VERSION,
"uscheduleconf DIR ACCT LOGACCT [JOBDIR [LOGDIR]]",
"This program creates an svscan service directory DIR, starting a task "
"to schedule for ACCT. Jobs will be read from JOBDIR or ~ACCT/.uschedule.\n"
"Logging information will be written to LOGDIR or ~ACCT/.uschedule/log, "
"using the account LOGACCT.\n"
"JOBDIR may be `-', in which case ~ACCT/.uschedule is used.\n",
"long",
"Report bugs to uschedule@lists.ohse.de",
4,6,0,0,uogetopt_out,myopts
};


int
main(int argc, char **argv)
{
  const char *username;
  const char *logusername;
  const char *jobdir=0;
  const char *logdir=0;
  static stralloc pw_dir, pw_shell, pw_name;

  bailout_progname(argv[0]);
  flag_bailout_fatal_begin=3;
  uogo_posixmode=1;
  myoptenv.program=flag_bailout_log_name;
  uogetopt_parse(&myoptenv,&argc,argv);

  dir=argv[1];
  username=argv[2];
  logusername=argv[3];
  if (argv[4]) {
    jobdir=argv[4];
    if (argv[5])
      logdir=argv[5];
  }

  /* band-aids: Code below doesn't quote >'<, and it's questionable
   * whether such quoting works correctly everywhere. Better play safe. */
#define BAND_AID(x) \
  if (x[str_chr(x,'\'')]) xbailout(100,0,#x " must not contain >'<: ",x,0,0);
  BAND_AID(dir);
  BAND_AID(username);
  BAND_AID(logusername);
  if (jobdir) BAND_AID(jobdir);
  if (logdir) BAND_AID(logdir);

  /* limit scope of pw. i often misused it below */
  {
    struct passwd *pw; 

    pw = getpwuid(getuid());
    /* setuidgid `random 0 65535` uscheduleconf ..., anyone? */
    if (!pw) xbailout(100,0,"failed to get current user data",0,0,0);

    pw = getpwnam(username);
    if (!pw) xbailout(100,0,"unknown account ",username,0,0);

    useruid=pw->pw_uid;
    usergid=pw->pw_gid;
    if (!stralloc_copys(&pw_shell,pw->pw_shell)) oom();
    if (!stralloc_0(&pw_shell)) oom();
    if (!stralloc_copys(&pw_dir,pw->pw_dir)) oom();
    if (!stralloc_0(&pw_dir)) oom();
    if (!stralloc_copys(&pw_name,pw->pw_name)) oom();
    if (!stralloc_0(&pw_name)) oom();

    pw = getpwnam(logusername);
    if (!pw) xbailout(100,0,"unknown account ",logusername,0,0);
    loguid=pw->pw_uid;
    loggid=pw->pw_gid;
  }

  if (!jobdir || str_equal(jobdir,"-")) {
    static stralloc d;
    if (!stralloc_copys(&d,pw_dir.s)) oom();
    if (!stralloc_cats(&d,"/")) oom();
    if (!stralloc_cats(&d,SCHEDULEDIR)) oom();
    if (!stralloc_0(&d)) oom();
    jobdir=d.s;
    BAND_AID(jobdir);
  }
  if (!logdir) {
    static stralloc d;
    if (!stralloc_copys(&d,jobdir)) oom();
    if (!stralloc_cats(&d,"/log")) oom();
    if (!stralloc_0(&d)) oom();
    logdir=d.s;
    BAND_AID(logdir);
  }
  if (logdir[str_chr(logdir,'\'')])
    /* band-aid */
    xbailout(100,0,"logdir must not contain >'<: ",logdir,0,0);

  /* root space */

  if (o_user_change)
    base_dir(dir,01700, rootuid,rootgid);
  else
    base_dir(dir,01700, useruid,usergid);

  start_file("run");
  outs("#! /bin/sh\n");
  outs("exec 2>&1\n");
  outs("cd '"); outs(jobdir); outs("' || exit 1\n");
  outs("exec \\\n");
  xsetuidgid(username);
  outs("./run\n");
  finish_file();
  xowner(rootuid,rootgid);
  perm(0700);

  mp("log",0700,rootuid,rootgid);
  start_file("log/run");
  outs("#! /bin/sh\n");
  outs("exec 2>&1\n");
  outs("cd '"); outs(logdir); outs("' || exit 1\n");
  outs("exec \\\n");
  outs("softlimit -m 8000000 -o 400 -p 40 \\\n");
  xsetuidgid(logusername);
  outs("./run\n");
  finish_file();
  xowner(rootuid,rootgid);
  perm(0700);

  /* user space */

  make_dir(jobdir);
  xowner(useruid,usergid);
  perm(0700);
  set_dir(jobdir);
  dir=jobdir;

  start_file("run");
  outs("#! /bin/sh\n");
  outs("exec \\\n");
  outs("envdir ./env \\\n");
  outs("uscheduled -d `pwd`\n");
  finish_file();
  xowner(useruid,usergid);
  perm(0700);
  make_fifo("fifo",0600);
  xowner(useruid,usergid);

  mp("env",0700,useruid,usergid);
  start_file("env/PATH");
  outs("/command:/usr/local/bin:/usr/bin:/bin");
  if (0==useruid)
    outs(":/usr/local/sbin:/usr/sbin:/sbin");
  outs("\n");
  finish_file();
  xowner(useruid,usergid);
  perm(0700);

  simple_env("env/HOME",pw_dir.s);
  simple_env("env/SHELL",pw_shell.s);
  simple_env("env/USER",pw_name.s);
  simple_env("env/LOGNAME",pw_name.s);


  mp(IDDIR,0700,useruid,usergid);
  make_dir(logdir);
  xowner(loguid,loggid);
  perm(0700);
  set_dir(logdir);
  {
    static stralloc logrun;
    if (!stralloc_copys(&logrun,logdir)) oom();
    if (!stralloc_cats(&logrun,"/run")) oom();
    if (!stralloc_0(&logrun)) oom();
    start_file(logrun.s);
    outs("#! /bin/sh\n");
    outs("exec \\\n");
    outs("softlimit -m 5000000 -o 100 -p 10 \\\n");
    outs("multilog t \\\n");
    outs("./\n");
    finish_file();
    xowner(loguid,loggid);
    perm(0700);
  }

  return 0;
}
